File: app/src/main/java/info/horriblesubs/sher/ui/e/Recent.java
package info.horriblesubs.sher.ui.e;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.view.View.OnClickListener;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.AppCompatTextView;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.DefaultItemAnimator;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import info.horriblesubs.sher.AppMe;
import info.horriblesubs.sher.R;
import info.horriblesubs.sher.adapter.ListItemAdapter;
import info.horriblesubs.sher.adapter.ListItemAdapter.OnItemClick;
import info.horriblesubs.sher.api.horrible.model.ListItem;
import info.horriblesubs.sher.api.horrible.response.Result;
import info.horriblesubs.sher.common.TaskListener;
import info.horriblesubs.sher.db.DataMethods;
import info.horriblesubs.sher.ui.a.Ads;
import info.horriblesubs.sher.ui.a.Menu;
import info.horriblesubs.sher.ui.a.SearchChange;
import info.horriblesubs.sher.ui.a.SearchChange.SearchListener;
import info.horriblesubs.sher.ui.c.Search;
import info.horriblesubs.sher.ui.i.Show;
import info.horriblesubs.sher.ui.z.LoadingDialog;

import static info.horriblesubs.sher.AppMe.appMe;

public class Recent extends AppCompatActivity implements TaskListener, OnItemClick,
        Observer<Result<ListItem>>, OnClickListener, SearchListener, SwipeRefreshLayout.OnRefreshListener {

    private SwipeRefreshLayout refreshLayout;
    private AppCompatTextView textView;
    private RecyclerView recyclerView;
    private LoadingDialog dialog;
    private Model model;
    private Menu menu;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        AppMe.appMe.setTheme();
        setContentView(R.layout.c_activity_0);
        model = new ViewModelProvider(this).get(Model.class);
        model.items(this).observe(this, this);
        refreshLayout = findViewById(R.id.swipeRefreshLayout);
        findViewById(R.id.fab).setOnClickListener(this);
        recyclerView = findViewById(R.id.recyclerView);
        new SearchChange(this, this);
        refreshLayout.setOnRefreshListener(this);
        textView = findViewById(R.id.textView);
        Ads.InterstitialAd.load(this);
        Ads.BannerAd.load(this);
        menu = Menu.all();
        if (!model.isExecuted()) model.onRefresh(false);
    }

    @Override
    public void onPostExecute() {
        refreshLayout.setRefreshing(false);
        if (dialog != null)
            dialog.dismiss();
    }

    @Override
    public void onPreExecute() {
        if (dialog == null)
            dialog = new LoadingDialog(this);
        dialog.show();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        model.onStopTask();
        onPostExecute();
        dialog = null;
    }

    @Override
    protected void onStart() {
        super.onStart();
        recyclerView.setLayoutManager(new GridLayoutManager(this, appMe.isPortrait()?1:2));
        recyclerView.setItemAnimator(new DefaultItemAnimator());
    }

    @Override
    public void onItemClicked(ListItem item) {
        if (item.link == null) return;
        Intent intent = new Intent(this, Show.class);
        intent.putExtra("show.link", item.link);
        startActivity(intent);
    }

    @Override
    public void onChanged(Result<ListItem> result) {
        if (result == null || result.items == null || result.items.size() == 0) {
            recyclerView.setVisibility(View.GONE);
            textView.setVisibility(View.VISIBLE);
            recyclerView.setAdapter(null);
        } else {
            recyclerView.setAdapter(ListItemAdapter.get(Recent.this, result.items));
            new DataMethods(this).onResetNotifications(result.items);
            recyclerView.setVisibility(View.VISIBLE);
            textView.setVisibility(View.GONE);
        }
    }

    @Override
    public void onTermChange(String s) {
        if (recyclerView.getAdapter() instanceof ListItemAdapter)
            ((ListItemAdapter) recyclerView.getAdapter()).onSearch(s);
    }

    @Override
    public void onSearch(String s) {
        if (s == null || s.length() < 2) {
            Toast.makeText(this, "Invalid Search Term", Toast.LENGTH_SHORT).show();
            return;
        }
        Intent intent = new Intent(this, Search.class);
        intent.putExtra("search", s);
        startActivity(intent);
    }

    @Override
    public void onClick(View view) {
        if (view.getId() == R.id.fab)
            menu.show(getSupportFragmentManager());
    }

    @Override
    public void onRefresh() {
        model.onRefresh(true);
    }
}

File: app/src/main/java/info/horriblesubs/sher/ui/main/explore/recent/Recent.kt
package info.horriblesubs.sher.ui.main.explore.recent

import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.PopupMenu
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.AppCompatTextView
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.DefaultItemAnimator
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import info.horriblesubs.sher.R
import info.horriblesubs.sher.adapter.ItemClick
import info.horriblesubs.sher.adapter.RecentAdapter
import info.horriblesubs.sher.api.horrible.model.ItemRecent
import info.horriblesubs.sher.api.horrible.result.Result
import info.horriblesubs.sher.common.Info
import info.horriblesubs.sher.common.LoadingListener
import info.horriblesubs.sher.common.inflate
import info.horriblesubs.sher.db.DataAccess
import info.horriblesubs.sher.dialog.InfoDialog
import info.horriblesubs.sher.dialog.LoadingDialog
import info.horriblesubs.sher.ui.show.Show

class Recent: Fragment(), LoadingListener, Observer<Result<ItemRecent>?>, ItemClick<ItemRecent>,
    PopupMenu.OnMenuItemClickListener {

    private val adapter: RecentAdapter = RecentAdapter(this)
    private var textView: AppCompatTextView? = null
    private var recyclerView: RecyclerView? = null
    private var dialog: LoadingDialog? = null
    private var vm: RecentVM? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        vm = ViewModelProvider(activity as AppCompatActivity).get(RecentVM::class.java)
        vm?.initialize(this)
    }

    override fun onCreateView(inflater: LayoutInflater, group: ViewGroup?, bundle: Bundle?): View? {
        return group?.inflate(R.layout._b_fragment_1_c, inflater)
    }

    override fun onViewCreated(view: View, bundle: Bundle?) {
        super.onViewCreated(view, bundle)
        textView = view.findViewById(R.id.textView2)
        recyclerView = view.findViewById(R.id.recyclerView)
        val info:InfoDialog? = context?.let{InfoDialog(it)}
        view.findViewById<View>(R.id.button).setOnClickListener{
            info?.show(Info.RECENT, vm?.result?.value)
        }
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        vm?.time?.observe(viewLifecycleOwner, Observer {textView?.text = it})
        vm?.result?.observe(viewLifecycleOwner, this)
        recyclerView?.layoutManager = LinearLayoutManager(context)
        recyclerView?.itemAnimator = DefaultItemAnimator()
        recyclerView?.adapter = adapter
        vm?.refresh()
    }

    override fun onChanged(result: Result<ItemRecent>?) {
        when {
            result == null || result.items.isNullOrEmpty() -> {
                Toast.makeText(context, "Server Error", Toast.LENGTH_SHORT).show()
                recyclerView?.visibility = View.GONE
                textView?.text = ("Never")
            }
            else -> {
                val list = result.items.toMutableList()
                recyclerView?.visibility = View.VISIBLE
                DataAccess.resetNotify(list)
                adapter.reset(list)
            }
        }
    }

    override fun stop() {
        dialog?.dismiss()
        dialog = null
    }

    override fun start() {
        if(dialog == null)
            dialog = context?.let{LoadingDialog(it)}
        dialog?.show()
    }

    override fun onDestroy() {
        super.onDestroy()
        vm?.stop()
        stop()
    }

    override fun onClick(e: ItemRecent?) {
        if (e == null) return
        val intent = Intent(activity, Show::class.java)
        intent.putExtra("link", e.link)
        startActivity(intent)
    }

    override fun onMenuItemClick(item: MenuItem?): Boolean {
        return when(item?.itemId) {
            R.id.toggle_b -> {
                adapter.toggle()
                true
            }
            R.id.refresh -> {
                vm?.refresh(true)
                true
            }
            else -> false
        }
    }
}